crates/noirc_printable_type/src/lib.rs

```diff
@@ -0,0 +1,199 @@
+                write!(fmt, "{name} {{ ")?;
+
+                let mut fields = fields.iter().peekable();
+                while let Some((key, field_type)) = fields.next()  {
+                    let value = &map[key];
+                    write!(fmt, "{key}: {}", PrintableValueDisplay::new(value, field_type))?;
+                    if fields.peek().is_some() {
+                        write!(fmt, ", ")?;
+                    }
+                }
+
+                write!(fmt, " }}")?;
+            }
+
+            _ => return Err(std::fmt::Error),
+        };
+        Ok(())
+    }
+}
+
+/// This trims any leading zeroes.
+/// A singular '0' will be prepended as well if the trimmed string has an odd length.
+/// A hex string's length needs to be even to decode into bytes, as two digits correspond to
+/// one byte.
+fn format_field_string(field: FieldElement) -> String {
+    if field.is_zero() {
+        return "0x00".to_owned();
+    }
+    let mut trimmed_field = field.to_hex().trim_start_matches('0').to_owned();
+    if trimmed_field.len() % 2 != 0 {
+        trimmed_field = "0".to_owned() + &trimmed_field;
+    }
+    "0x".to_owned() + &trimmed_field
+}
+
+// TODO: Figure out a better API for this to avoid exporting the function
+/// Assumes that `field_iterator` contains enough [FieldElement] in order to decode the [PrintableType]
+pub fn decode_value(
+    field_iterator: &mut impl Iterator<Item = FieldElement>,
+    typ: &PrintableType,
+) -> PrintableValue {
+    match typ {
+        PrintableType::Field
+        | PrintableType::SignedInteger { .. }
+        | PrintableType::UnsignedInteger { .. }
+        | PrintableType::Boolean => {
+            let field_element = field_iterator.next().unwrap();
+
+            PrintableValue::Field(field_element)
+        }
+        PrintableType::Array { length, typ } => {
+            let length = *length as usize;
+            let mut array_elements = Vec::with_capacity(length);
+            for _ in 0..length {
+                array_elements.push(decode_value(field_iterator, typ));
+            }
+
+            PrintableValue::Vec(array_elements)
+        }
+        PrintableType::String { length } => {
+            let field_elements: Vec<FieldElement> = field_iterator.take(*length as usize).collect();
+
+            PrintableValue::String(decode_string_value(&field_elements))
+        }
+        PrintableType::Struct { fields, .. } => {
+            let mut struct_map = BTreeMap::new();
+
+            for (field_key, param_type) in fields {
+                let field_value = decode_value(field_iterator, param_type);
+
+                struct_map.insert(field_key.to_owned(), field_value);
+            }
+
+            PrintableValue::Struct(struct_map)
+        }
+    }
+}
+
+// TODO: Figure out a better API for this to avoid exporting the function
+pub fn decode_string_value(field_elements: &[FieldElement]) -> String {
```

Comment:
I believe this is a repeat as we still have a `decode_string_value` in `noirc_abi`? Could we switch to using one or the other? Or is this part of the TODO at the top of this method?

Reply:
This crate and `noirc_abi` cannot use functions from each other so they don't cause a circular dependency, so this function and the function above are currently duplicated in each crate but they probably shouldn't be public at and instead should be implemented as `From` or serde.